### ISSUE TITLE
TypeScript 4.7 Iteration Plan

### ISSUE THREAD

**Issue by DanielRosenwasser** [microsoft/TypeScript#48027]
This document outlines our focused tasks for TypeScript 4.7. It minimally indicates intent to investigate tasks or contribute to an implementation. Nothing is set in stone, but we will strive to complete these tasks in a reasonable timeframe.

Date           | Event
---------------|-------------------------
February 22nd  | TypeScript 4.6 Release
April 1st      | Create 4.7 Beta (4.7.0) Build for Testing
April 5th      | **TypeScript 4.7 Beta Release**
May 6th        | Create 4.7 RC (4.7.1) Build for Testing
May 10th       | **TypeScript 4.7 RC Release**
May 20th       | Create 4.7 Final (4.7.2) Build for Testing
May 24th       | **TypeScript 4.7 Final Release** 🚀

## Language and Compiler Features

* Node ESM UX Issues, and Move Node ESM Mode from Nightly
* [Instantiation Expressions](https://github.com/microsoft/TypeScript/pull/47607)
* [Import Strategies for `import type` and `/// <reference />`](https://github.com/microsoft/TypeScript/pull/47807)
* [Investigate Contextually Typed And Compatible Operator](https://github.com/microsoft/TypeScript/issues/47920)
* [`lib.d.ts` Updates](https://github.com/microsoft/TypeScript/issues/48018)

## Editor Productivity

* [Object Member Snippet Completions](https://github.com/microsoft/TypeScript/issues/46590)
* [Investigate Go to JavaScript Implementation](https://github.com/microsoft/TypeScript/issues/6209)
* [Prototype Virtual File System Support in TSServer](https://github.com/microsoft/TypeScript/issues/47600)
* [Prototype Automatic Type Acquisition in Web Contexts](https://github.com/microsoft/TypeScript/issues/45314)
* [Easy Run/Debug for Loose TypeScript Files](https://github.com/microsoft/vscode/issues/124280)
* [Investigate TS Server Plugins on the Web](https://github.com/microsoft/TypeScript/issues/47376)

## Performance

* [Investigate Outstanding Performance Issues](https://github.com/microsoft/TypeScript/issues?q=is%3Aopen+is%3Aissue+label%3A%22Domain%3A+Performance%22)
* [Find All References Refactoring and Optimizations](https://github.com/microsoft/TypeScript/pull/47938)
* [Supporting Additional Trace Points](https://github.com/microsoft/TypeScript/issues/40460)
* [Profiling/Tracing Mode Improvements](https://github.com/microsoft/TypeScript/issues/40461)
* [Investigate Tooling for JIT Deoptimizations](https://github.com/microsoft/TypeScript/issues/44211)

## Infrastructure

* [Migrate to Modules](https://github.com/microsoft/TypeScript/pull/35561)
* Determine Shipping Strategy for Modules
* [Fuzzer Improvements and Addressing Crashes](https://github.com/microsoft/TypeScript/issues/44210)
* [Improving User Suite Reports](https://github.com/microsoft/TypeScript/issues/44209)
* [Expand Perf Suite](https://github.com/microsoft/TypeScript/issues/44033)

## Website

* [Investigate Playground as a VS Code Extension](https://github.com/microsoft/TypeScript-Website/issues/2156)

**Comment by vietanhtran16** [comment]
Hello, I am not sure if this is the right place for this but we got the below issue with version `4.7.0-dev.20220320` (this is automatically selected by VSCode) like the below image. We did some testing and noticed that this issue does not show up for version below `4.7.0`, in our case `4.6.2` and `4.1.5` work fine. If this is not the right place for this, please let me know. Thanks

Context: 
- We provide type safe fetch (its type is quite nested) which allows us to safely type the response based on the url of the request we are trying to make for a good dev experience
- We are getting the below error when we use our fetch in a `Promise.all`. My assumption is that Promise.all
 is trying to infer the type from our fetch type which is quite nested and caused this error on `4.7.0-dev.20220320`

```
const [responseOne, responseTwo] = await Promise.all([ // it complains about the below error here
    fetch("https://something/v1", { req, timeout: 5000 }),
    fetch(
      "https://something/v2",
      { req, timeout: 5000 }
    )
  ])
```
 ![image](https://user-images.githubusercontent.com/23372284/159402681-4381527d-8fc4-4027-8deb-63e933509547.png)

**Comment by aladdin-add** [comment]
@vietanhtran16  yes, the issue is not for bug report, please file a new issue. :)

**Comment by vietanhtran16** [comment]
> @vietanhtran16 yes, the issue is not for bug report, please file a new issue. :)

Cool I will file a new one

**Comment by DanielRosenwasser** [comment]
@typescript-bot create release-4.7

**Comment by DanielRosenwasser** [comment]
Looks like it's probably failing because of the Node ESM stuff which we were hoping to unflag for the beta at the least. There's also some basic bug fixes I'd like us to pull in before beta, so maybe we'll go out on 4/6 instead.

**Comment by DanielRosenwasser** [comment]
@typescript-bot create release-4.7

**Comment by DanielRosenwasser** [comment]
[The feature roadmap](https://github.com/microsoft/TypeScript/wiki/Roadmap#47-may-2022) now reflects what we've been able to get done for 4.7.

**Comment by Jack-Works** [comment]
I still want to suggest typescript to drop Node 12 in new Node ESM. Lets start from Node 14. Node 14 has TLA support.

![image](https://user-images.githubusercontent.com/5390719/162572695-a8d8f788-cd2d-416b-91dc-3a91ed54cfa8.png)

Node 12 will end LTS this month.

**Comment by lvelden** [comment]
Hi! I've seen the [4.7-beta announcement](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-beta), but it's not listed under [releases](https://github.com/microsoft/TypeScript/releases) and there's also no [v4.7-beta tag](https://github.com/microsoft/TypeScript/tags). Are you planning on adding this?

**Comment by DanielRosenwasser** [comment]
Sorry, beta tag should be out https://github.com/microsoft/TypeScript/releases/tag/v4.7-beta

**Comment by diervo** [comment]
@DanielRosenwasser is there a specific tag or semantics for filing issues against this beta that will help streamline the process?

**Comment by magnattic** [comment]
Not sure if this is a TS or yarn issue and I didn't want to cross-post, but just for reference, 4.7 seems to [break when used with yarn pnp mode](https://github.com/yarnpkg/berry/issues/4385). If it's a TS problem I can also open an issue here.

**Comment by DanielRosenwasser** [comment]
Just open up issues and fill out the template, maybe signal specifically whether you've found a regression or a bug in a new feature.

**Comment by DanielRosenwasser** [comment]
We've removed `resolution-mode` from `import type` and `export type`, and moved it into nightly-only, so I cannot summon the bot. The bump to RC is at https://github.com/microsoft/TypeScript/pull/49007.

**Comment by DanielRosenwasser** [comment]
4.8 Iteration Plan is available at https://github.com/microsoft/TypeScript/issues/49074

**Comment by RyanCavanaugh** [comment]
@typescript-bot bump release-4.7

**Comment by Jamesernator** [comment]
Just wanted to point out one mistake in the [TypeScript 4.7 RC release notes](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7-rc/#ecmascript-module-support-in-node-js) that should probably be fixed for the full release and that's that the line:

```
Certain global-like values like require() and process cannot be used directly.
```

Is actually incorrect for `process`, unlike `require` which a per-cjs-module injected thing, the `process` object is actually just a property of `global` so is usable in ESM as a global without problems.

An alternative object for the sentence could the [`module`](https://nodejs.org/api/modules.html#the-module-object) per-cjs-module object i.e.:

```
Certain global-like values like require() and module cannot be used directly.
```

**Comment by DanielRosenwasser** [comment]
Thank you for that correction!

**Comment by DanielRosenwasser** [comment]
[TypeScript 4.7 is now out!](https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/)

**Comment by HolgerJeromin** [comment]
Happy to test, but linked [Typescript SDK](https://marketplace.visualstudio.com/items?itemName=TypeScriptTeam.TypeScript-47) has a 404 and [nuget.org](https://www.nuget.org/packages/Microsoft.TypeScript.MSBuild) has a `4.7.2-rc` (hu?).
But [npm](https://www.npmjs.com/package/typescript) is available :-)

**Comment by Dens2017** [comment]
I installed the typescript 4.7.2 but I encountered an error

Module build failed (from ./node_modules/ts-loader/index.js):
Error: Debug Failure. False expression: Non-string value passed to `ts.resolveTypeReferenceDirective`, likely by a wrapping package working with an outdated `resolveTypeReferenceDirectives` signature. This is probably not a problem in TS itself.

I'm using Node@16.15.0 and npm@8.11.0

Here are some of the packages I used
@fluentui/react@8.70.0
@types/react@17.0.45
react@17.0.2
react-dom@7.0.2
@types/node@16.11.36
@typescript-eslint/eslint-plugin@4.29.0
@typescript-eslint/parser@4.29.0
eslint@7.32.0
eslint-config-standard@16.0.2
eslint-plugin-import@2.26.0
eslint-plugin-node@11.1.0
eslint-plugin-promise@5.1.0
typescript@4.7.2

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-4.7

**Comment by DanielRosenwasser** [comment]
@typescript-bot bump release-4.7
